WordPress/gutenberg
# Query loop / Post template: Enable post format filter

## What?

Enables filtering the query loop result by post format.
Partial for https://github.com/WordPress/gutenberg/issues/53049

Additional props @justintadlock @SergeyBiryukov 

## Why?

This feature is often requested by users, see the linked issue.

## How?

Description updated September 5 2024.

This PR attempts to enable filtering the post query by post format, but handles the post format **separately from
other taxonomies**, as requested by reviewers.

### REST API changes:
A new class `Gutenberg_REST_Posts_Controller_6_7` that extends `WP_REST_Posts_Controller` is added to the 6.7 compatibility folder.
The intention of this new temporary class is to add 'format' to the Posts controller, so that it can be used to filter the query result per post format.

Updates are made to these class methods: `get_item`, `get_collection_params`

### Query loop block:
- Adds a new parameter, `format`, to the `query` block attribute in the query loop block.
- Adds a new `FormatControls` that uses a list of supported post formats inside a `FormTokenField`.
- `getThemeSupports` is used to get the supported post formats from the store.

### Post Template block:
- The new `format` parameter is passed from the query loop block to the post template block as part of the `query`.
- The post format is added to the `query` using the new `format` key (as opposed to using the post format taxonomy).

### Filtering query_loop_block_query_vars
Filter the query vars (tax_query) for the Query Loop block to support post formats on the front of the site.

## Todo & follow ups
_To be handled as a follow up in a separate PR:_
Ensure that the query loop displays the correct posts when the block is set to **inherit** the query, when the block is placed in a post format archive template.


## Testing Instructions
Enable post formats in a theme by adding the theme support.  Example:
```
function twentytwentyfour_setup(){
	add_theme_support( 'post-formats', array( 'standard', 'aside', 'gallery', 'audio', 'video', 'link', 'image', 'chat', 'status', 'quote' ) );
}

add_action( 'after_setup_theme', 'twentytwentyfour_setup' );
```
Add some test content, both with and without post formats assigned.

Test the query loop both in a template and in a post/page.

Add or select an existing query loop block.
Enable the **custom** query type in the query loop settings sidebar.

Find and open the Filters panel.
Select the three dot menu on the filters panel and toggle all available filters.
Confirm that one of the filters is an input field for "Formats.".
Confirm that all post formats that you registered with the theme supports can be entered into and selected in the input field.

Test that there are no regressions with the query loop features and its filters:
- When the format filter is not enabled
- When the format filter is enabled
- When a single format is selected
-  When multiple formats are selected
- Especially test that the 'standard' format displays posts that has no post format assigned, when it is chosen as the only format, and when combined with another format.
- Confirm that you can reset the format filter with the reset options in the Filter panel.

- Test with classic themes and block themes
- Test that there are no regressions if post formats are not enabled on the theme (All features should work as before this patch).

### Testing custom post types
Follow the steps above to register support for post formats in the theme.

Register a custom post type with post format support, include:
```
'show_in_rest' => true 
'supports' => array( 'title', 'editor', 'thumbnail', 'post-formats'),
```

```
function twentytwentyfour_custom_post_type() {
	register_post_type( 'test',
		array(
			'labels' => array(
				'name' => __( 'Tests' ),
				'singular_name' => __( 'Test' ),
			),
			'public' => true,
			'has_archive' => true,
			'show_in_rest' => true,
			'supports' => array( 'title', 'editor', 'thumbnail', 'post-formats'),
			'taxonomies' => array( 'category', 'post_tag' ),
			'rewrite' => array('slug' => 'tests'),
		)
	);
}

add_action( 'init', 'twentytwentyfour_custom_post_type' );
```

In the admin, create a few items of this post type, with and without a post format assigned.
Go to Appearance > Editor and select a template with a query -or create one.

1. In the query loop setting, change query type to custom.
2. Under post type, select your post type.
3. Open the Filters panel. Confirm that you can toggle the display of the Formats control.
4. When you select / focus on / place the cursor on the input for the formats control there should be a dropdown with the same formats that you registered with the theme support.
5. Confirm that entering a random value that is not a registered format does nothing.
6. Select a format. Confirm if the query loop displays the correct items in the editor and front.
7. Select multiple formats. Confirm if the query loop displays the correct items in the editor and front.

### Switching post types (After registering your custom post type)
In the settings sidebar of the query loop, set the post type to posts.
Open the Filters panel, and add a post format.
Switch the post type to the custom post type. Confirm that the filter is unchanged.
Switch the post type to page, which does not support post formats: Confirm that the filter is not showing in the panel.
Switch the post type to posts: confirm that the filter is reset (it was reset when you switched to the page).

### Testing the query pagination
After creating a number of posts with formats, set the "items per page" setting to something low so that the query pagination block will show on the front.
Save and view the front of the site. Use the pagination and confirm that page 2 etc displays the correct posts.

## Screenshots or screencast 
I apologise, I have compressed the video but it is still almost 4 MB.
In this video I am testing different post formats in a template in the Site Editor.
I am testing formats in combination with other taxonomies, entering a non-existing format name in the control, etc.
I am switching between Post, page, and a custom post type with post format support, to illustrate that the filter is reset when switching to pages, but maintained when switching between posts and custom post types.

(Aside: I am not using a mouse and I do see that I often miss click when trying to remove a format from the control)

https://github.com/user-attachments/assets/9dc8597b-9956-4f13-aa91-bf0030484fa1
